### PR TITLE
fix(rust, python): don't panic on `str.parse_int`

### PR DESCRIPTION
**Pull request by sorhawell**
I suggest to upgrade parse_int to not panic! on parseError/overflow or Null inputs.
Also i32 might be a little small for many realistic use cases so I changed to i64.

A future PR could be to include a strict param to not silently return Null on parseError

original function author: @cojmeister

### CODE REVIEW

**Comment by ritchie46** [comment]
Thanks @sorhawell. Can you split this one up? One with the fix and then another PR that changes the output type as that is a breaking change.

The breaking PR can then be merged on a later time.

**Comment by sorhawell** [comment]
@ritchie46  Ok I will divert i64 from this PR
Would it make sense to immediately add defaulted arg strict=TRUE to raise error on any parse/overflow error. Then the behavior change is only the error message, unless user set strict=FALSE.

**Comment by ritchie46** [comment]
> @ritchie46 Ok I will divert i64 from this PR Would it make sense to immediately add defaulted arg strict=TRUE to raise error on any parse/overflow error? Then the behavior change is only the error message, unless user set strict=FALSE.

Yeap, that's good. :+1:

File: polars/polars-ops/src/chunked_array/strings/namespace.rs [hunk 1]
```diff
@@ -63,11 +63,24 @@ pub trait Utf8NameSpaceImpl: AsUtf8 {
...
-        let f = |s: &str| <i32 as Num>::from_str_radix(s, radix.unwrap_or(2)).unwrap();
-        ca.apply_cast_numeric(f)
+        if strict {
+            let f = |opt_s: Option<&str>| -> PolarsResult<Option<i32>> {
+                opt_s
+                    .map(|s| <i32 as Num>::from_str_radix(s, radix))
+                    .transpose()
+                    .map_err(|err| PolarsError::ComputeError(err.to_string().into()))
+            };
+            let ca: PolarsResult<Int32Chunked> = ca.into_iter().map(f).collect();
```
Comment (sorhawell) [inline polars/polars-ops/src/chunked_array/strings/namespace.rs:76]: Is there a way to do collect_trusted() on a PolarsResult<T>?

Comment (ritchie46) [reply polars/polars-ops/src/chunked_array/strings/namespace.rs:76]: Nope, I don't think so. Could investigate that iterator adaptor later.

File: polars/polars-arrow/src/utils.rs [hunk 2]
```diff
@@ -154,16 +154,6 @@ impl<T: NativeType> FromTrustedLenIterator<Option<T>> for PrimitiveArray<T> {
     }
 }
 
-impl<T: NativeType> FromTrustedLenIterator<T> for PrimitiveArray<T> {
```
Comment (ritchie46) [inline polars/polars-arrow/src/utils.rs:157]: Why is this removed?

File: polars/polars-ops/src/chunked_array/strings/namespace.rs [hunk 3]
```diff
@@ -63,11 +63,24 @@ pub trait Utf8NameSpaceImpl: AsUtf8 {
...
     // Parse a string number with base _radix_ into a decimal (i32)
-    fn parse_int(&self, radix: Option<u32>) -> Int32Chunked {
+    fn parse_int(&self, radix: u32, strict: bool) -> PolarsResult<Int32Chunked> {
+        use polars_arrow::utils::CustomIterTools;
         let ca = self.as_utf8();
-
-        let f = |s: &str| <i32 as Num>::from_str_radix(s, radix.unwrap_or(2)).unwrap();
-        ca.apply_cast_numeric(f)
+        if strict {
+            let f = |opt_s: Option<&str>| -> PolarsResult<Option<i32>> {
```
Comment (ritchie46) [inline polars/polars-ops/src/chunked_array/strings/namespace.rs:70]: We could maybe speed this up and take the same branch as `non_strict` and then simply check the null_count afterwards. If the null_count differs from `self`, strict should fail.

File: polars/polars-arrow/src/utils.rs (hunk 2 above)
Comment (sorhawell) [reply polars/polars-arrow/src/utils.rs:157]: I'm so sorry that was a mistake. I was trying to for some time to implement trusted collect for Result next to there, but deleted it again. I did not the check the diff well enough before committing.

File: polars/polars-ops/src/chunked_array/strings/namespace.rs (hunk 3 above)
Comment (sorhawell) [reply polars/polars-ops/src/chunked_array/strings/namespace.rs:70]: Uhh that is smart. I will implement that.

**Comment by cojmeister** [comment]
Hi @sorhawell thanks for making my contribuition better!
I'm trying to implement the opposite function now - to_string_radix but I don't know where to start. 
Should I make it in: 
`polars-lazy>polars-plan>rc>dsl>arithmetic.rs>impl Expr`?

**Comment by sorhawell** [comment]
@cojmeister I mainly work with py-polars and r-polars so I'm somewhat new to this also. I neither think it should be in string subnamespace at least. It will try to look into it on Monday.

File: polars/polars-ops/src/chunked_array/strings/namespace.rs (hunk 1 above)
Comment (sorhawell) [reply polars/polars-ops/src/chunked_array/strings/namespace.rs]: not needed anymore

File: polars/polars-ops/src/chunked_array/strings/namespace.rs [hunk 4]
```diff
@@ -63,11 +63,37 @@ pub trait Utf8NameSpaceImpl: AsUtf8 {
...
+        use polars_arrow::utils::CustomIterTools;
         let ca = self.as_utf8();
+        let f = |opt_s: Option<&str>| -> Option<i32> {
+            opt_s.and_then(|s| <i32 as Num>::from_str_radix(s, radix).ok())
+        };
+        let out: Int32Chunked = ca.into_iter().map(f).collect_trusted();
+
+        if strict && ca.null_count() != out.null_count() {
+            let failure_mask = !ca.is_null() & out.is_null();
+            let failures = ca.filter(&failure_mask)?.unique()?;
```
Comment (sorhawell) [inline polars/polars-ops/src/chunked_array/strings/namespace.rs:76]: I was missing a unique_stable() method for ChunkedArrays to improve the err msg, I guess it could be implemented analogous to arg_unique in another PR

**Comment by sorhawell** [comment]
@cojmeister 
I suppose I would try an implementation analogous to top_k_impl<T> where T: PolarsNumericType and everywhere else top_k is mentioned.

**Review by ritchie46 — COMMENTED** [review]
Sorry, i forgot to finish a review so it was still pending. Almost ready, one remark.

File: polars/polars-ops/src/chunked_array/strings/namespace.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (sorhawell) [reply polars/polars-ops/src/chunked_array/strings/namespace.rs]: I have sliced, and sorted. Added a count of failures in err msg also. It turns out the fmt_list() it self caps at 4 items.  But I guess it is still good to slice before sorting.
```py
>>> import polars as pl
>>> s = pl.Series("bin", ["110", "101", "010", "1","bob","a","b","c","d","e","f","g","h","i","j"])
>>> s.str.parse_int(2, True)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/sorenwelling/Documents/projs/polars/py-polars/polars/internals/series/utils.py", line 98, in wrapper
    return s.to_frame().select(f(*args, **kwargs)).to_series()
  File "/Users/sorenwelling/Documents/projs/polars/py-polars/polars/internals/dataframe/frame.py", line 5980, in select
    self.lazy()
  File "/Users/sorenwelling/Documents/projs/polars/py-polars/polars/internals/lazyframe/frame.py", line 1323, in collect
    return pli.wrap_df(ldf.collect())
exceptions.ComputeError: Strict integer parsing failed for 11 value(s): ["a", "b", ... "j"] ParseIntError msg for first shown value: 'invalid digit found in string'. Consider non strict parsing.
```

**Comment by ritchie46** [comment]
Alright. Looks good. Thanks a lot @sorhawell
